### PR TITLE
Introduce C# 8 and start using Nullable Reference Types

### PR DESCRIPTION
**Pull request by drewnoakes**
Apologies for this large PR out of nowhere. I started it mostly out of curiosity and in the end it seemed like something the team might like to consider or at least discuss. There may be legitimate reasons we can't move to C# 8, but now that it's generally available with VS2019 version 16.3, I am personally all for moving to it, especially for Nullable Reference Types.

## Proposed changes

- Introduce C# 8
- Convert two projects to SDK-style
- Apply several code fixes suggested by the newer compiler
  - 74dacf33b09d42196e5a9ed1ea2424bb50b42ff3 Use static local functions
  - 26fb3d15bb8ed52d087d1845da407cc7da9ba5e7 Use compound assignment
  - a6274e7ce8d14a2159b7baaa390e9e7f91b71cd9 Use readonly fields
  - 6f6257bf8f8c41ac922e489fbc22f4e154bd1b97 Use simple using statements
  - 62c4847535463f4902b8b9143c8746bb9fc1e412 Use object initializers
  - 4160c6a602d71133a11b7e77946894b94f649785 Remove redundant assignments
  - ae8162e122cd97640a5466ee28214b41146f6db2 Inline variable declarations
  - 0e0c39a18fa782311b8709a1503ced6fc2687edc Use switch expressions
  - 2f33c4bc0d570d3927355fdc37dd3d8ef82edba3 Simplify collection initialization
- Manual C# 8 changes
  - c13253fb6d10c6be065055704b8f695016195960 Use readonly members on readonly structs
  - db8bf49681e94fe78f5bd278ecfcf118a006c421 `Annotate GitUIPluginInterfaces` for Nullable Reference Types
  - 309aa5fa66015a100f7a0260a2186885058634af Annotate `GitExtUtils` for Nullable Reference Types

## Test methodology 

- Manual validation of changes and testing
- Running unit tests

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by drewnoakes** [comment]
I'm not sure why [this test](https://ci.appveyor.com/project/gitextensions/gitextensions/builds/28177069#L274) failed on CI. It passes locally. Have pushed another commit, so will see if it's just flakey.

> Merging #7321 into master will decrease coverage by 11.44%.

This seems suspicious! Maybe the tests didn't complete, or some tests weren't included.

It's late here, I'll look more another day.

**Comment by sharwell** [comment]
I'll start looking through these commits

File: GitCommands/Git/Tag/GitCreateTagCmd.cs [hunk 1]
```diff
@@ -34,28 +34,19 @@ protected override ArgumentString BuildArguments()
 
             string GetArgumentForOperation()
             {
-                switch (CreateTagArguments.Operation)
+                return CreateTagArguments.Operation switch
                 {
                     /* Lightweight */
```
Comment (sharwell) [inline GitCommands/Git/Tag/GitCreateTagCmd.cs]: 💭 Well these comments seem unnecessary 😂

File: GitCommands/Git/ConflictData.cs [hunk 2]
```diff
@@ -15,10 +15,10 @@ namespace GitCommands.Git
...
 
-        public ConflictedFileData Base { get; }
-        public ConflictedFileData Local { get; }
-        public ConflictedFileData Remote { get; }
+        public readonly ConflictedFileData Base { get; }
+        public readonly ConflictedFileData Local { get; }
+        public readonly ConflictedFileData Remote { get; }
 
-        public string Filename => Local.Filename ?? Base.Filename ?? Remote.Filename;
+        public readonly string Filename => Local.Filename ?? Base.Filename ?? Remote.Filename;
```
Comment (sharwell) [inline GitCommands/Git/ConflictData.cs]: 😕 Why are we doing this? `readonly struct` by definition has only `readonly` members.

File: Plugins/Bitbucket/CreatePullRequestRequest.cs [hunk 3]
```diff
@@ -39,24 +39,30 @@ protected override JObject ParseResponse(JObject json)
...
+            var resource = new JObject
+            {
+                ["title"] = _info.Title,
+                ["description"] = _info.Description,
 
-            resource["fromRef"] = CreatePullRequestRef(
+                ["fromRef"] = CreatePullRequestRef(
                 _info.SourceRepo.ProjectKey,
-                _info.SourceRepo.RepoName, _info.SourceBranch);
+                _info.SourceRepo.RepoName, _info.SourceBranch),
```
Comment (sharwell) [inline Plugins/Bitbucket/CreatePullRequestRequest.cs]: :bulb: Indent one more

File: Plugins/Bitbucket/CreatePullRequestRequest.cs [hunk 4]
```diff
@@ -39,24 +39,30 @@ protected override JObject ParseResponse(JObject json)
...
 
-            resource["toRef"] = CreatePullRequestRef(
+                ["toRef"] = CreatePullRequestRef(
                 _info.TargetRepo.ProjectKey,
-                _info.TargetRepo.RepoName, _info.TargetBranch);
+                _info.TargetRepo.RepoName, _info.TargetBranch)
```
Comment (sharwell) [inline Plugins/Bitbucket/CreatePullRequestRequest.cs]: :bulb: Indent one more

File: UnitTests/GitCommandsTests/ArgumentBuilderExtensionsTests.cs [hunk 5]
```diff
@@ -263,7 +263,7 @@ public void Add_ForcePushOptions_ensure_all_switches_are_handled()
             foreach (int mode in Enum.GetValues(typeof(ForcePushOptions)))
             {
                 // unimplemented switches will result in InvalidEnumArgumentException
-                var args = new ArgumentBuilder
+                new ArgumentBuilder
```
Comment (sharwell) [inline UnitTests/GitCommandsTests/ArgumentBuilderExtensionsTests.cs]: 💡 Consider using a discard so it's clear to the reader that this is not accidental

```
_ = new ArgumentBuilder
```

File: GitExtUtils/GitUI/TaskExtensions.cs [hunk 6]
```diff
@@ -1,200 +0,0 @@
-﻿using System;
-using System.Runtime.CompilerServices;
-using System.Threading;
-using System.Threading.Tasks;
-using Microsoft;
-
-namespace GitUI
-{
-    public static class TaskExtensions
```
Comment (sharwell) [inline GitExtUtils/GitUI/TaskExtensions.cs]: ❓ Why is this deleted? Did we upgrade vs-threading to use the one defined there?

File: GitCommands/Statistics/CommitCounter.cs [hunk 7]
```diff
@@ -1,75 +0,0 @@
-﻿using System;
-using System.Collections.Generic;
-using GitUIPluginInterfaces;
-
-namespace GitCommands.Statistics
-{
-    public static class CommitCounter
```
Comment (sharwell) [inline GitCommands/Statistics/CommitCounter.cs]: ❓ Why was this deleted as part of moving to SDK project?

File: GitExtUtils/GitUI/ThreadHelper.cs [hunk 8]
```diff
@@ -16,11 +17,11 @@ public static class ThreadHelper
...
 
-        private static JoinableTaskContext _joinableTaskContext;
-        private static JoinableTaskCollection _joinableTaskCollection;
-        private static JoinableTaskFactory _joinableTaskFactory;
+        private static JoinableTaskContext? _joinableTaskContext;
+        private static JoinableTaskCollection? _joinableTaskCollection;
+        private static JoinableTaskFactory? _joinableTaskFactory;
 
-        public static JoinableTaskContext JoinableTaskContext
+        public static JoinableTaskContext? JoinableTaskContext
```
Comment (sharwell) [inline GitExtUtils/GitUI/ThreadHelper.cs]: 💭 It's a bit strange for this to be nullable but `JoinableTaskFactory` to not be. Do we ever check the return value of this for null? If not, consider changing the getter to:

```
return _joinableTaskContext ?? throw new InvalidOperationException("vs-threading is not configured")`
```

Since the setter will still need to _accept_ null, you can use one of the following approaches to indicate that the property can be assigned null but will never return it:

1. Declare the property as nullable, but apply `[NotNull]` to it
2. Declare the property as non-nullable, but apply `[AllowNull]` to it

File: GitExtUtils/GitUI/ThreadHelper.cs [hunk 9]
```diff
@@ -49,7 +50,7 @@ internal set
             }
         }
 
-        public static JoinableTaskFactory JoinableTaskFactory => _joinableTaskFactory;
+        public static JoinableTaskFactory JoinableTaskFactory => _joinableTaskFactory!;
```
Comment (sharwell) [inline GitExtUtils/GitUI/ThreadHelper.cs]: :bulb: Consider using `?? throw new InvalidOperationException(...)` instead of `!`

File: GitExtUtils/GitUI/ThreadHelper.cs [hunk 10]
```diff
@@ -58,7 +59,7 @@ public static void ThrowIfNotOnUIThread([CallerMemberName] string callerMemberNa
                 return;
             }
 
-            if (!JoinableTaskContext.IsOnMainThread)
+            if (!JoinableTaskContext!.IsOnMainThread)
```
Comment (sharwell) [inline GitExtUtils/GitUI/ThreadHelper.cs]: 💡 All of the `!` lines for this property can be removed if the property is changed to not return null.

**Comment by drewnoakes** [comment]
Thanks for the review Sam 👍. I'll go through your comments soon.

Before that though I'd like to hear from other team members if there is anything blocking migrating to VS2019/C# 8.

**Comment by gerhardol** [comment]
I do not have any objection. If it is a step forward for you, it is OK for me.
#5087 is enough for me.

What stopped us earlier to upgrade to VS2019 (only) was WIX toolset not prepared and Appveyor not prepared.
There will likely be sporadic users questioning this move though.

**Comment by drewnoakes** [comment]
> #5087 is enough for me.

I will never forget that PR number...

> What stopped us earlier to upgrade to VS2019 (only) was WIX toolset not prepared and Appveyor not prepared.

Good point. I noticed the installer project did not load, which didn't surprise me as I hadn't installed the tooling. We should validate the installer project works before merging any of this.

AppVeyor appears to be ok with VS2019 now, based on this PR.

**Comment by mast-eu** [comment]
>> What stopped us earlier to upgrade to VS2019 (only) was WIX toolset not prepared and Appveyor not prepared.

> We should validate the installer project works before merging any of this.

I took the branch `drewnoakes:nullable-reference-types` (241521bcd1e55c7798d50572b4459a886597b8fa) and locally merged it into `gitextensions:release/3.2` (bc07663c3c893f1d896520e0bdcfbe35cfe4b729). In other words, I merged the current state of this PR into the latest release of GE.

There were only a few minor merge conflicts. Nothing caused by this PR though, but rather by existing differences between `gitextensions:master`  and `gitextensions:release/*` . Anyhow, nothing problematic.

The result of my local merge has been pushed to `mast-eu:release/test_drew` (c2b0d8953f642918ffa113041c9164260c877979).
I picked my own fork, because I wanted to avoid "noise" in `gitextensions` repo and already had AppVeyor configured for my fork. 
I renamed the branch, because AppVeyor does different things on branches called `release/*`.

As a result, I got the expected build artifacts: https://ci.appveyor.com/api/buildjobs/djybk7fr77aapqbk/artifacts/combined.3.3.0.180-unsigned.zip
Here's the log: https://ci.appveyor.com/project/mast-eu/gitextensions/builds/28202782

The ZIP contains both Portable and MSI builds. These are the same files that @RussKie would use for a release. The MSI installs without problems on my Win10. 
I never tried to build a MSI locally on my PC. No idea if this is working or not. But building on AppVeyor, with VS 2019 (v16.3.4), worked out of the box.

**Comment by RussKie** [comment]
Awesome!
I'll find sometime to look at the changes next week.

**Comment by drewnoakes** [comment]
I'm going to close this PR as after talking to @RussKie it sounds like it'd be a while before this would be merged and I don't have the bandwidth to keep this up to date. I'm happy for others to cherry-pick from it (as in #7396).

My goal was to start using nullable reference types in this repo, as I believe it adds a lot of value. This requires a few steps:

- Upgrade projects to SDK-style
- Update lang version to C# 8
- Start annotating types

This is probably too much for a single PR.

**Comment by drewnoakes** [comment]
>  I believe c13253f is unnecessary and can be reverted.

Thanks @sharwell, you're right. This would make a nice analyzer.

File: Plugins/GitUIPluginInterfaces/CredentialsManager.cs [hunk 11]
```diff
@@ -44,35 +43,40 @@ public void Save()
             }
         }
 
-        protected NetworkCredential GetCredentialOrDefault(SettingLevel settingLevel, [NotNull] string name, NetworkCredential defaultValue)
+        protected NetworkCredential GetCredentialOrDefault(SettingLevel settingLevel, string name, NetworkCredential defaultValue)
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply Plugins/GitUIPluginInterfaces/CredentialsManager.cs]: An interesting idea, though I'm not sure I've needed this in my annotation efforts to date. I'll consider this for future, especially when swapping out `JetBrains.Annotations` attributes, possibly in tandem with `<Nullable>annotations</Nullable>`.

File: Plugins/GitUIPluginInterfaces/CredentialsManager.cs [hunk 12]
```diff
@@ -44,35 +43,40 @@ public void Save()
...
         {
             var targetName = GetWindowsCredentialsTarget(name, settingLevel);
             if (string.IsNullOrWhiteSpace(targetName))
             {
                 return defaultValue;
             }
 
-            if (Credentials.TryGetValue(targetName, out var result) || AdysTechCredentialManagerWrapper.TryGetCredentials(targetName, out result))
+            if (Credentials.TryGetValue(targetName!, out var result) || AdysTechCredentialManagerWrapper.TryGetCredentials(targetName!, out result))
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply Plugins/GitUIPluginInterfaces/CredentialsManager.cs]: In the project system we used a similar [`Strings`](https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed/Utilities/Strings.cs) class for this.

File: Plugins/GitUIPluginInterfaces/GitUIPluginInterfaces.csproj [hunk 13]
```diff
@@ -49,93 +20,35 @@
...
-    <ProjectReference Include="..\..\GitExtUtils\GitExtUtils.csproj">
-      <Project>{0f1f1168-a4b2-4fa2-b17b-735140d17f39}</Project>
-      <Name>GitExtUtils</Name>
-    </ProjectReference>
-  </ItemGroup>
+  <!--
   <ItemGroup>
     <EmbeddedResource Include="UserControls\CredentialsControl.resx">
       <DependentUpon>CredentialsControl.cs</DependentUpon>
     </EmbeddedResource>
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply Plugins/GitUIPluginInterfaces/GitUIPluginInterfaces.csproj]: Should have been removed.

[the remark this answers is not in the document]
Comment (drewnoakes) [reply Plugins/GitUIPluginInterfaces/GitUIPluginInterfaces.csproj]: I think it should have been removed, or possibly reinstated.

File: Plugins/GitUIPluginInterfaces/IGitModule.cs [hunk 14]
```diff
@@ -72,9 +70,7 @@ public interface IGitModule
         /// <summary>Indicates HEAD is not pointing to a branch (i.e. it is detached).</summary>
         bool IsDetachedHead();
 
-        [ContractAnnotation("=>false,objectId:null")]
-        [ContractAnnotation("=>true,objectId:notnull")]
-        bool TryResolvePartialCommitId(string objectIdPrefix, out ObjectId objectId);
+        bool TryResolvePartialCommitId(string objectIdPrefix, [NotNullWhen(returnValue: true)] out ObjectId objectId);
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply Plugins/GitUIPluginInterfaces/IGitModule.cs]: Yes, should have been `ObjectId?`.

File: GitExtUtils/GitUI/TaskExtensions.cs (hunk 6 above)
Comment (drewnoakes) [reply GitExtUtils/GitUI/TaskExtensions.cs]: This (and `CommitCounter.cs`) were not included in the csproj files before, and were brought in my globbing. The unused files were removed.

Comment (RussKie) [reply GitExtUtils/GitUI/TaskExtensions.cs]: I've done the same in my two other PRs.
